Importing/exporting config data --> Master

harishbommakanti opened the pull request:
Features added to config page to allow importing and exporting season data like elements/pit scouting questions.

harishbommakanti committed:
Added local jquery file

- Local jquery file so user device doesn't have problems when not connected to internet
- removed some commented code in teams.html

harishbommakanti committed:
Template for exporting data

- added importing-Exporting-Config.js
-  added code to be a basic template of the export button, need to add functionality now

harishbommakanti committed:
Download link works

- pressing the link will download JSON of config info into a json into a file
- need to work on styling the button and making the json file styled for user ease

harishbommakanti committed:
Styling for export link

- added styling for link to look like a button
- need to work on styling the json file in a user friendly format

harishbommakanti committed:
Export file is 'prettified'

- JSON data in downloaded file is 'prettified', meaning it has tabs and is more friendly to users
- Exporting config data is good?

harishbommakanti committed:
Added html/js template for import button

- base code for importing the season config off of an external json file

harishbommakanti committed:
Importing config data works

- User can input a .json file and import the configuration from the file, overwriting/adding to prev saved data

mshafer1 reviewed (changes requested):
I would suggest moving import/export to its own page.

Also, don't you think events and teams to scout would also be something to have the lead scout setup then export and share with the team? (i.e., would it make sense to have all of the season config, event config, and team config all be part of the export and import file)?

I definitely think it would be annoying if I have to add the season (by case sensitive string) then import some of the configuration for it.

It would work better if you moved the export and import to it's own page, on export, ask which seasons to include (maybe check boxes), exported all data related to selected seasons. - you could add an entry to your dictionary that's something like `"Season Keys": [{season 1}, {season 2}]`, so that you can know what seasons to look for on import.

On import, it is reasonable to expect that the user wants to blop over any seasons they already have (maybe the scouting lead sent out a new version of the pit questions), but we should think about possibly having a confirmation like "this config file contains data that will overwrite your configuration for [{season(s)}], continue?"

Side note, just give the file a custom extension (maybe something like `.rcubedscoutconfig`) and assume if you get that file extension that it's valid (make this a constant and reference it by name 😄 )  - if you make a comment about making sure it's valid JSON, now you're asking people who may not be coders or even familiar JSON to know something, but if you give them a file type for export and import, that makes it clear what this is for.

(didn't look at code, as I anticipate this discussion will prompt some major changes).

harishbommakanti committed:
PR fix: imp/exp data is in its own page

- title
- added directory changes to the navbar
- changed text on index.html

harishbommakanti committed:
PR fix: added season checkbox

- Added a checkbox for seasons to let user indicate which seasons to export/import data for
- added js code for ^ which sets a cookie
- need to work on reading the different file type (.rouse...) and working with the curr checked seasons

harishbommakanti committed:
PR fix: buttons able to work with checkbox

- functionality works with checkboxes
- now, need to go in and add the season elements and team data to the whole output object (for exporting, worry about importing later)

harishbommakanti committed:
PR fix: can now export events of a season

- exporting the data can now also report the events stored under a specific season
- now, need to work on looping through the local storage and adding that huge json string as another kv pair in the output object for exporting

harishbommakanti committed:
PR fix: exporting team data

- user can now see team data in the export file
- export file now is 'config.rcubedscoutconfig' instead of 'config.json'
- some method renaming

harishbommakanti committed:
PR fix: importing event list, new seasons

- user can now import the event list from the given file
- also, if new seasons are added, those will be added to the seasonList cookie (as the first kv pair of the file is 'seasons in this config')

mshafer1 reviewed (changes requested):
Conceptually, this layout is better, but when I went to export data, I just got a blank file.
It's throwing a JSON parse error under:
```
VM856:1 Uncaught SyntaxError: Unexpected end of JSON input
    at JSON.parse (<anonymous>)
    at returnTeamDataArr (importing-Exporting-Config.js:153)
    at getExportData (importing-Exporting-Config.js:124)
    at exportData (importing-Exporting-Config.js:52)
    at HTMLAnchorElement.<anonymous> (importing-Exporting-Config.js:319)
    at HTMLAnchorElement.dispatch (jquery.js:4)
    at HTMLAnchorElement.v.handle (jquery.js:4)
```

This is my current cookie string:
```
/season_config/Deep Space/Ball=10; /season_config/Deep Space/Hatch=15; _ga=GA1.1.1024030647.1572104924; Deep Space events=test1Δtest2Δ; /event_data teams/Deep Space/test1={"123":"test1","456":"test2","1234":"test3"}; events=; Deep Space pitQuestions=test?Ω; currCheckedTeam=; currCheckedEvent=/Deep Space/test2; seasonList=Deep Space§[object Object]§; currCheckedSeason=[object Object]; /imp-exp-data/checkedSeasons=Deep Space; arp_scroll_position=140
```

and my local storage:
```
> JSON.stringify(Object.entries(localStorage)
[["/Deep Space/test1/1234/Match Data/28","[{\"Match #\":\"28\"},{\"Ball\":\"8\"},{\"Hatch\":\"10\"}]"],["/Deep Space/test1/123/pit","[{\"test?\":\"test\"}]"],["/Deep Space/test1/123/Match Data/27","[{\"Match #\":\"27\"},{\"Ball\":\"17\"},{\"Hatch\":\"5\"}]"],["/Deep Space/test1/456/Match Data/27","[{\"Match #\":\"27\"},{\"Ball\":\"15\"},{\"Hatch\":\"7\"}]"]]
```

harishbommakanti committed:
PR fix: Exporting JSON-parse error fixed

- exporting file is now (hopefully) bug free

harishbommakanti commented:
@mshafer1 I think that error is fixed on my end. Also, the importing data button should be working to import the event data and team data along with the config data.

harishbommakanti committed:
PR fix: Importing Data button works

- Importing data feature also sets up team data and event data cookies added on to the pre-PR config-data only

harishbommakanti committed:
PR fix: Added an import sesaon modal, moved some UI elements around

- added a modal that would remind users that the input file would overwrite the seasons that they specify
- moved the checkboxes to be in the export region rather than the import region

mshafer1 commented on scripts/importing-Exporting-Config.js:
```diff
@@ -0,0 +1,508 @@
+{
+    //set a cookie for checkedSeasons
+    var arr = [];
+    $('input[type=checkbox]').each(function () 
```
this will run on all check boxes, might want to do `$('#seasonRadioList-imp-exp-data input[type=checkbox]')` to get inputs of type checkbox under the seasonRadioList

mshafer1 commented on scripts/importing-Exporting-Config.js:
```diff
@@ -0,0 +1,508 @@
+
+function getExportData() //returns prettified (tabbed) json.stringify version of current config
+{
+    var checkedSeasonList = getCurrCheckedSeasons().split(','); // csv string --> array
```
Why are you routing this through a cookie?

If I reload the page, are you going to reset the check boxes based on the cookie?

If not, it would be simpler (and more efficient on memory long term) to just take the array of values and return them to your variable here.

mshafer1 commented on scripts/importing-Exporting-Config.js:
```diff
@@ -0,0 +1,508 @@
+    var checkedSeasonList = getCurrCheckedSeasons().split(','); // csv string --> array
+
+    var output = {};
+    output['Seasons in this configuration'] = checkedSeasonList; //sets season list at top of config
```
this is kind of a long name. And you have to have the EXACT SAME CASE AND SPACING on the import function.

Put it in a constant so that it is easy to tell that it is the same string.
```js
var const EXPORT_CONFI_SEASON_LIST_KEY  = '...'
```

mshafer1 commented on scripts/importing-Exporting-Config.js:
```diff
@@ -0,0 +1,508 @@
+                    var currEvent = eventList[k];
+                    eventsArr.push(currEvent);
+                }
+            }
```
This is essentially the same code twice over.
Put it in a function.

Take advantage of the fact that JS passes objects by reference copy (that is, all objects are referred to by copy in JS, so when you pass one to a function, you pass a copy of the reference).

This allows you to pass to the function:
* the cookie,
* the desired start
* You CAN pass the seperator, but this raises questions why they are different.
* the array to append to.

(and the function basically becomes the if statement).

that would leave with something like:
```js
try_parse_from_cookie(cookie, pitStart, questionsArr);
try_parse_from_cookie(cookie, eventStart, eventsArr);
```

mshafer1 commented on scripts/importing-Exporting-Config.js:
```diff
@@ -0,0 +1,508 @@
+        }
+
+        // looping through cookies for each season \/ \/
+        var elementsObjName = season + ' elements';
```
and what if season contains ' elements' in it? or ' events'?

Just a thought, but would it be beneficial to store in a slightly different format like:
```json
{
  "{Season 1}":{
   "elements": elementsArr,
  "pit Questions": questionsArr,
  "Events": eventsArr,
  "TeamData": returnTeamDataArr(...)
  },
  "{Season 2}":{
  ...
 }
}
```

mshafer1 commented on scripts/importing-Exporting-Config.js:
```diff
@@ -0,0 +1,508 @@
+        var content = e.target.result;
+        var obj = JSON.parse(content);
+
+        if (file.name.endsWith('.rcubedscoutconfig'))
```
And what do you do if it does not match this condition?

Consider flipping these `if`s around:
```js
if(!file.name.endsWith(config_file_name_extension)) {
  return; // or whatever error handling you want
}

var continueOrNot = getImportModalChoice();
...
```

mshafer1 commented on scripts/importing-Exporting-Config.js:
```diff
@@ -0,0 +1,508 @@
+function processImportData(obj) //goes through inner objects and assigns values to cookies
+{
+    var keys = Object.keys(obj);
+    var seasonList = obj[keys[0]]; //seasons at top of import config file
```
this assumes the seasons list is at the top. JSON is not considered to be ordered, and different browsers may or may not include them in that order. Better to re-use the constant from my earlier comment.

mshafer1 commented on scripts/importing-Exporting-Config.js:
```diff
@@ -0,0 +1,508 @@
+            if (currObj.startsWith(elementObjName)) //add elements/points cookies
+            {
+                //set element value pairs
+                var innerArr = obj[currObj];
```
you know the exact dictionary keys by name.

go straight to
```js
var elementsObj = obj[eventObjName];
var pitObj = obj[pitObjName];
...
```

mshafer1 commented on scripts/importing-Exporting-Config.js:
```diff
@@ -0,0 +1,508 @@
+                }
+            }
+    
+            if (currObj.startsWith(pitObjName)) //add pitQuestions cookie
```
see previous comment about using names/keys for direct access.

mshafer1 commented on scripts/importing-Exporting-Config.js:
```diff
@@ -0,0 +1,508 @@
+                setCookie(cookieName,cookieValString,750);
+            }
+
+            if (currObj.startsWith(eventObjName)) //add eventList cookie for the season
```
see previous comment about using names/keys for direct access.

mshafer1 commented on importing-exporting-data.html:
```diff
@@ -0,0 +1,79 @@
+        <div>
+            <h1 id = "title">FRC 6321 Scouting</h1>
+        </div>
+        <div id = "mySidenav" class="sidenav">
+            <a href="javascript:void(0)" class="closebtn"
+            onclick = "closeNav()">&times;</a>
+            <a href="index.html">Home</a>
+            <a href="season-config.html">Season Configuration</a>
+            <a href="event-data.html">Event Data</a>
+            <a href="team-data.html">Team Data</a>
+            <a href="importing-exporting-data.html">Importing and Exporting Data</a>
+        </div>
```
I have a suspicion this `nav` isn't updated everywhere.

After this PR, while waiting to get the demo with the team done, look into using Jekyll to include this top level stuff. (I can link you to the docs and some examples if desired).

harishbommakanti replied:
Yeah, I've noticed that too: when I go to the Team Data page the text of the nav bar button reverts to times new roman. I remember you talking me through how to do this on the slack a few months ago but further docs and links would be appreciated.

mshafer1 replied:
Further docs:
https://jekyllrb.com/docs/

and links:
https://jekyllrb.com/
https://github.com/mshafer1/themepuzzler

Jekyll is the tool that gh-pages uses to build your site anyways.
If you setup with an `_includes` folder and put `nav.html` in there,
then each of your pages can have
`{% include nav.html %}`
to have the contents of `nav.html` put in at that point (kind of like a sub vi, but in-lined when the site is "built" into static pages).

Jekyll is a very powerful tool built for static websites and blogs - but even just having a basic layout (see the docs) or a few files you can include (like the nav), goes a long ways towards making things that should be updated just once but affect everywhere.

mshafer1 commented on scripts/importing-Exporting-Config.js:
```diff
@@ -0,0 +1,508 @@
+    var seasonList = object[keys[0]]; // season config list is the entry for the first key
+    
+    //add seasons to modal text
+    var modalText = 'This file contains data that will overwrite the following season(s): ';
```
put a `<br/>` at the end of this line, maybe between each season name, and after them all.

Whitespace helps people to parse quickly. After the first time I've used your app, I expect to see a list of seasons, if I don't remember what's in this config file, I might want to pull this up and quickly skim the seasons in the file - which will be easier if they're not in the middle of a paragraph.

mshafer1 reviewed (changes requested):
Seems to work pretty well.

I might argue that the team data is not really configuration, but. .. this does mean we have a way of conjoining all least some different team data (as they would be different teams) without additional work, so I would suggest leaving it for now.

harishbommakanti committed:
PR fix: deleted cookies for checked seasons, modified input[] for getCurrCheckedSeason()

- getCurrCheckedSeason() no longer references a cookie, just returns an array
- jquery input selector modified

harishbommakanti committed:
PR fix: parse_cookie() added, redid file reading error handling,

- added a parse_cookie() method that simplifies the exporting pit questions and event questions methods

harishbommakanti committed:
PR fix: added seasonListKey constant, removed importFile redundancy

- added a constant to make it easier to read the season list key
- removed an extra for loop in the processImport method as specific keys are already known

harishbommakanti committed:
PR fix: added <br>'s to modal text

- added <br>'s to modal text to make it easier for users to see which seasons the import file with run through

harishbommakanti committed:
PR fix: modified exporting JSON format

changed it to:
{
  "{Season 1}":{
   "elements": elementsArr,
  "pit Questions": questionsArr,
  "Events": eventsArr,
  "TeamData": returnTeamDataArr(...)
  },
  "{Season 2}":{
  ...
 }
}

harishbommakanti committed:
PR fix: made JSON format easier to process for Exporting

- instead of having arrays inside objects inside arrays and stuff like that, most of the json object is just based around objects
- fixed exporting, now I will work on improving the importing to work with this new scheme

harishbommakanti committed:
PR fix: Importing data also works with new JSON scheme

- deleted a random setCheckedSeasons cookie in the load function
- importing function now works with the hopefully simpler JSON scheme

harishbommakanti committed:
Fixed some Team Data table flaws

- When testing the imp/exp features, the pit table was showing data for ALL events, so I fixed that
- made it that you could see the match data onload of the page instead of having to click on a team

mshafer1 commented on importing-exporting-data.html:
```diff
@@ -0,0 +1,79 @@
+<script src="scripts/eventDataScript.js"></script> <!--js script-->
+<script src="scripts/teamDataScript.js"></script> <!--js script-->
+<script src="scripts/commonMethodHelpers.js"></script> <!--js script-->
+<script src="scripts/importing-Exporting-Config.js"></script> <!--js script-->
```
I just caught this - these really belong inside the `<head>` section.

mshafer1 commented on scripts/importing-Exporting-Config.js:
```diff
@@ -0,0 +1,480 @@
+
+            //set up match/pit objects
+            //match_Obj[match_Obj_Name] = match_Obj_Value;
+            //pit_Obj[pit_Obj_Name] = pit_Obj_Value;
```
don't comment out code. Delete it and make a commit with an explanative message why.

mshafer1 commented on scripts/importing-Exporting-Config.js:
```diff
@@ -0,0 +1,480 @@
+        var content = e.target.result;
+        var obj = JSON.parse(content);
+
+        if (!file.name.endsWith('.rcubedscoutconfig'))
```
move this before trying to `JSON.parse` the contents.

mshafer1 commented on scripts/importing-Exporting-Config.js:
```diff
@@ -0,0 +1,480 @@
+        var continueOrNot = getImportModalChoice();
+        if (continueOrNot.includes('Yes')) //user clicked yes
+        {
+            closeImportModal();
```
the `close` should be part of the getting.

mshafer1 commented on scripts/importing-Exporting-Config.js:
```diff
@@ -0,0 +1,480 @@
+        {
+            var curr_Team_Key = teams_In_Curr_Event[j];
+            var curr_Team_Num = curr_Team_Key.split(':')[0];
+            var curr_Team_Name = curr_Team_Key.split(':')[1];
```
what if the team name has a `:` in it?

might be better to key off of team numbers and have `"name": "..."` be one of the values you store (in addition to the match data and pit data).

mshafer1 commented on scripts/importing-Exporting-Config.js:
```diff
@@ -0,0 +1,480 @@
+        } else
+        {
+            modalText += `${seasonList[i]},<br/> `;
+        }
```
90% of the code is the same between the if and the else. Make the part that's different obvious (ideally by having it be the only part of the if statement). Might look like:

```js
modalText += seasonList[i];
if(i == seasonList.lenth - 1) {
  modelText += '.';
} else {
 modalText += ',';
}

modalText += '<br/>';
```

When you break it down like that, it looks like you could even use a ternary operator (which is similar to the LabVIEW select).

```js
modalText += (i == seasonList.length - 1)?'.':',';
```

mshafer1 reviewed (approved):
Looking good.

harishbommakanti committed:
PR fix: added scripts to <head>, removed commented code for imp-exp-data.js

- removed commented code that wasn't needed as a new JSON scheme is there

harishbommakanti committed:
PR fix: added close as part of getImportModalChoice

- close() added to getImportData()
- file reader checks if a file is .rouserobotics... before parsing the json

harishbommakanti committed:
PR fix: simplified if-else for modal text code

- simplified the modal text code using a ternary js operator

harishbommakanti committed:
PR fix: changed team naming scheme

- team name is now a kv pair alongside match/pit data: {'name' : 'Rouse'}

harishbommakanti committed:
PR fix: moved importing all js scripts to the <head> section for all html pages
